backend/src/api/OAuth2Controller.ts

```diff
@@ -0,0 +1,340 @@
+      windowMs: 60 * 60 * 1000 /* 1 hour */,
+      max: 10,
+      skipSuccessfulRequests: true,
+      standardHeaders: false,
+      legacyHeaders: false
+    });
+
+    const commonLimiter = rateLimit({
+        windowMs: 60 * 1000 /* 1 minute */,
+        max: 60,
+        standardHeaders: false,
+        legacyHeaders: false,
+        keyGenerator: (req) => String(req.session.data?.userId)
+    });
+
+    this.router.post('/oauth2/clients', commonLimiter, validate(listClientsSchema), (req, res) => this.listClients(req, res));
+    this.router.post('/oauth2/client', commonLimiter, validate(getClientSchema), (req, res) => this.getClientByClientId(req, res));
+    this.router.post('/oauth2/client/register', registerLimiter, validate(clientRegisterSchema), (req, res) => this.register(req, res));
+    this.router.post('/oauth2/client/regenerate-secret', commonLimiter, validate(clientManageSchema), (req, res) => this.regenerateClientSecret(req, res));
+    this.router.post('/oauth2/client/update-logo', commonLimiter, validate(updateLogoUrlSchema), (req, res) => this.updateClientLogoUrl(req, res));
+    this.router.post('/oauth2/client/delete', commonLimiter, validate(clientManageSchema), (req, res) => this.deleteClient(req, res));
+    this.router.post('/oauth2/client/change-visibility', commonLimiter, validate(clientManageSchema), (req, res) => this.changeClientVisibility(req, res));
+    this.router.post('/oauth2/authorize', commonLimiter, (req, res) => this.oauthExpressServer.authorize({
+      authenticateHandler: {
+        handle: async (req) => {
+          await req.session.restore(req.body['X-Session-Id']);
+          if (req.session.isBarmalini() && req.session.getAgeMillis() > /*1 hour*/ 60 * 60 * 1000) {
+            await req.session.destroy();
+            return null;
+          }
+          if (!req?.session?.data?.userId) {
+            return null;
+          }
+          return {
+            id: req.session.data.userId
+          };
+        }
+      }
+    })(req, res, () => {}));
+    this.router.post('/oauth2/unauthorize', commonLimiter, validate(clientManageSchema), (req, res) => this.unAuthorizeClient(req, res));
+    this.router.post('/oauth2/token', commonLimiter, validate(getTokenSchema), (req, res) => this.oauthExpressServer.token({})(req, res, () => {}));
+  }
+
+  /**
+   * Create (register) new OAuth2 client
+   */
+  async register(request: APIRequest<OAuth2RegisterRequest>, response: APIResponse<OAuth2RegisterResponse>) {
+    if (!request.session.data.userId) {
+      return response.authRequired();
+    }
+
+    try {
+      const {name, description, logoUrl, initialAuthorizationUrl, redirectUris, isPublic} = request.body;
+      const userId = request.session.data.userId;
+      const author = await this.userManager.getById(userId);
+
+      const client: OAuth2ClientRaw = await this.oauth2Manager.registerClient(name, description, logoUrl, initialAuthorizationUrl, redirectUris, userId, isPublic);
```

Comment:
let's limit the total number of created clients per user (say, 32 is enough) to avoid abuse